### ISSUE TITLE
Add hooks for more mouse events

### ISSUE THREAD

**Issue by daniel-santos** [wesnoth/wesnoth#7949]
### Describe the desired feature

Hello,

I would love to be able to hook the right-click event, prior to the pop-up menu displaying so I could modify a menu entry based upon context. However, it looks like I'll have to do that in mouse-over, which will be more CPU wasteful. All-in-all, it would be nice to just have access to a greater number of them and let the add-on designer just not write crappy code that would lock up the game.

**Comment by CelticMinstrel** [comment]
Does a right-click trigger the existing mouse pressed event? I don't remember if I've checked that.

**Comment by gfgtdf** [comment]
> Does a right-click trigger the existing mouse pressed event? I don't remember if I've checked that.

Pretty sure it doesnt, The mostclick event is iirc implemented via one of the "select/move/attack" ,  "Move/Attack" Or  "Select" hotkeys, and right-click menu bypass the hotkey system iirc.

**Comment by doofus-01** [comment]
Is filtering the context for the various menu items to show up not enough?  Can you post an example?

**Comment by CelticMinstrel** [comment]
It's not enough when the text needs to be different for every hex and can dynamically change from turn to turn.

I heard that variables in menu descriptions are also not substituted before the menu is shown, so perhaps that should be opened as a separate issue. If the menu item name contains `$turn_number` (which would be written in the WML as `$|turn_number` to defer substitution) then it should show the current turn number each time you open the menu.

**Comment by doofus-01** [comment]
> It's not enough when the text needs to be different for every hex and can dynamically change from turn to turn.

The hexes can be certainly be filtered.  I've never tried for turns but it should be possible, especially if #7948 is fixed.

So, still need an example.

**Comment by CelticMinstrel** [comment]
#7948 is what I referenced in the second paragraph of my previous comment.

I suppose it's not impossible to add one menu item for every hex, but that seems like a kinda crazy workaround for this and #7948…

**Comment by doofus-01** [comment]
Crazy workaround for what?  Example, please.

**Comment by gfgtdf** [comment]
> However, it looks like I'll have to do that in mouse-over, which will be more CPU wasteful

I want to point out that there are a lot this things happening on "mouse over" already, in particular all the calculations for the right side panel, in particular the attack stats, like checking the resistance of all attacks against all enemy unit types, and checking whether abilities are active.

So depending on your concrete usecase it's quite possible that wasting cpu is is much less of an issue than you think it is, especially once https://github.com/wesnoth/wesnoth/issues/7948 is implemented.

**Comment by CelticMinstrel** [comment]
For what it's worth, this came up after a discussion on IRC in #wesnoth-umc-dev, so if you're impatient and don't want to wait for the original poster to explain, you can go look it up in the logs.

**Comment by daniel-santos** [comment]
Hello all. I didn't check my email, so I hadn't noticed this discussion! My apologies!

My use case is in Wesband where I would like to have a top-level menu item to pick up loot. Currently, in order to "get all items", you have to go in to several layers of menus and then back out again.

Wesband has a concept of "in-combat" and "out of combat" (or safe and not safe). If the unit is more than a certain distance from any enemy unit and hasn't attacked or used a "complex action", then after you move, the moves are refreshed and you can continue this until you use a non-looting action or come near hostile units.

So when not in combat, I would like to have a more liberal "get the loot" option that will pick up all gold and items under the unit and on adjacent hexes, and when in combat to still support the more conservative option, which is to loot just the hex (iirc, that consumes a "simple action").

I could just do that with two separate menu options and use the `[show_if]` filter to decide which to show, but I would also prefer to let the player know *what* they are about to pick up. So when in combat and only gold, it reads "Pick up _n_ gold" and when out of combat, "Pick up _n_ gold nearby." When only items it reads  "Pick up _n_ items" and "Pick up _n_ items nearby", respectively and when there's both  "Pick up _n_ gold and _n_ items" and "Pick up _n_ gold and  _n_ items nearby".

**Comment by doofus-01** [comment]
> For what it's worth, this came up after a discussion on IRC in #wesnoth-umc-dev, so if you're impatient and don't want to wait for the original poster to explain, you can go look it up in the logs.

My impatience was with you pretending to answer anything, not with the OP.

>So when not in combat, I would like to have a more liberal "get the loot" option that will pick up all gold and items under the unit and on adjacent hexes, and when in combat to still support the more conservative option, which is to loot just the hex (iirc, that consumes a "simple action").

>I could just do that with two separate menu options and use the [show_if] filter to decide which to show, but I would also prefer to let the player know what they are about to pick up. So when in combat and only gold, it reads "Pick up n gold" and when out of combat, "Pick up n gold nearby." When only items it reads "Pick up n items" and "Pick up n items nearby", respectively and when there's both "Pick up n gold and n items" and "Pick up n gold and n items nearby".

At its core, is the issue that you need to treat a far away target hex as if the unit is in a different state (in-combat/out-of-combat) than it currently is?  It sounds like something that should be possible with filtering, but you've probably found a detail that makes it not work.  I can look into this next weekend - I'm not a programmer but I am familiar with WML.

**Comment by gfgtdf** [comment]
> I could just do that with two separate menu options and use the [show_if] filter to decide which to show, but I would also prefer to let the player know what they are about to pick up. So when in combat and only gold, it reads "Pick up n gold" and when out of combat, "Pick up n gold nearby." When only items it reads "Pick up n items" and "Pick up n items nearby", respectively and when there's both "Pick up n gold and n items" and "Pick up n gold and n items nearby".

Except for printing the exact number of items/gold this can probably indeed be implemented by having ~6 different menu items one for each of the different cases that you mentioned.

**Comment by daniel-santos** [comment]
> At its core, is the issue that you need to treat a far away target hex as if the unit is in a different state (in-combat/out-of-combat) than it currently is?

No, the menu only comes up when right-clicking on a unit, but I want to allow the player to pick up loot from all of the adjacent tiles without visiting each of them, but only if they aren't "in combat". I could have two separate menus with filters that decides which will show (Wesband has an `[is_safe]` filter for that), but I would also like to put in the context menu entry description a summary of what they are about to pick up. I had also considered letting them know if the level of any of the items were "vendor trash" or "you might want this item", but decided against it.

**Comment by daniel-santos** [comment]
Hrm, I mentioned this issue in the PR, but it isn't linking here. So this is the PR: https://github.com/wesnoth/wesnoth/pull/7952

**Comment by Pentarctagon** [comment]
Addressed by #7952
